move/its/sources/its.move

```diff
@@ -1,287 +1,558 @@
 ) {
-    self.relayer_discovery_id = object::id(relayer_discovery);
-}
+    let value = self.value_mut!(b"receive_interchain_transfer");
+    
+    let (_, payload) = decode_approved_message(value, approved_message);
+    let mut reader = abi::new_reader(payload);
+    assert!(
+        reader.read_u256() == MESSAGE_TYPE_INTERCHAIN_TRANSFER,
+        EInvalidMessageType,
+    );
+
+    let token_id = token_id::from_u256(reader.read_u256());
+    reader.skip_slot(); // skip source_address
+    let destination_address = address::from_bytes(reader.read_bytes());
+    let amount = reader.read_u256();
+    let data = reader.read_bytes();
+
+    assert!(data.is_empty(), EInterchainTransferHasData);
 
-public(package) fun relayer_discovery_id(self: &ITS): ID {
-    self.relayer_discovery_id
+    let coin = value
+        .coin_management_mut(token_id)
+        .give_coin<T>(amount, clock, ctx);
+
+    transfer::public_transfer(coin, destination_address)
 }
 
-public(package) fun set_trusted_address(
+public fun receive_interchain_transfer_with_data<T>(
     self: &mut ITS,
-    chain_name: String,
-    trusted_address: String,
-) {
-    self.address_tracker.set_trusted_address(chain_name, trusted_address);
+    approved_message: ApprovedMessage,
+    channel: &Channel,
+    clock: &Clock,
+    ctx: &mut TxContext,
+): (String, vector<u8>, vector<u8>, Coin<T>) {
+    let value = self.value_mut!(b"receive_interchain_transfer_with_data");
+
+    let (source_chain, payload) = decode_approved_message(
+        value,
+        approved_message,
+    );
+    let mut reader = abi::new_reader(payload);
+    assert!(
+        reader.read_u256() == MESSAGE_TYPE_INTERCHAIN_TRANSFER,
+        EInvalidMessageType,
+    );
+
+    let token_id = token_id::from_u256(reader.read_u256());
+
+    let source_address = reader.read_bytes();
+    let destination_address = reader.read_bytes();
+    let amount = reader.read_u256();
+    let data = reader.read_bytes();
+
+    assert!(
+        address::from_bytes(destination_address) == channel.to_address(),
+        EWrongDestination,
+    );
+    assert!(!data.is_empty(), EInterchainTransferHasNoData);
+
+    let coin = value.coin_management_mut(token_id).give_coin(amount, clock, ctx);
+
+    (source_chain, source_address, data, coin)
 }
 
-public(package) fun set_trusted_addresses(
+public fun receive_deploy_interchain_token<T>(
     self: &mut ITS,
-    trusted_addresses: TrustedAddresses,
+    approved_message: ApprovedMessage,
 ) {
-    let (mut chain_names, mut trusted_addresses) = trusted_addresses.destroy();
-    let length = chain_names.length();
-    let mut i = 0;
-    while (i < length) {
-        self.set_trusted_address(
-            ascii::string(chain_names.pop_back()),
-            ascii::string(trusted_addresses.pop_back()),
-        );
-        i = i + 1;
-    }
+    let value = self.value_mut!(b"receive_deploy_interchain_token");
+
+    let (_, payload) = decode_approved_message(value, approved_message);
```

Comment:
we're differentiating from the gateway here by keeping most of the logic in this contract vs the versioned module. Shall we move it there for consistency? While the coupling is relatively lower, it still feels a bit awkward to keep them separate?

Reply:
I like it better this way because the storage logic lives in `its_v0` but the public facing logic live in `its`. I can change it if you think it's too awkward.